dataplane/updater.h

```diff
@@ -0,0 +1,219 @@
+	{
+	}
+
+	eResult update(const std::vector<common::acl::tree_chunk_8bit_t>& values)
+	{
+		stats.extended_chunks_size = std::max((uint64_t)object_type::extended_chunks_size_min,
+		                                      values.size() / 16);
+
+		for (;;)
+		{
+			pointer = memory_manager->create<object_type>(name.data(),
+			                                              socket_id,
+			                                              object_type::calculate_sizeof(stats.extended_chunks_size));
+			if (pointer == nullptr)
+			{
+				return eResult::errorAllocatingMemory;
+			}
+
+			eResult result = pointer->fill(stats, values);
+			if (result != eResult::success)
+			{
+				/// try again
+				memory_manager->destroy(pointer);
+				stats.extended_chunks_size *= 2;
+				continue;
+			}
+
+			break;
+		}
+
+		return eResult::success;
+	}
+
+	void clear()
+	{
+		if (pointer)
+		{
+			memory_manager->destroy(pointer);
+			pointer = nullptr;
+		}
+	}
+
+	void limits(common::limit::limits& limits) const
+	{
+		limits.emplace_back(name + ".extended_chunks",
+		                    socket_id,
+		                    stats.extended_chunks_count,
+		                    stats.extended_chunks_size);
+	}
+
+	void report(nlohmann::json& report) const
+	{
+		report["pointer"] = to_hex(pointer);
+		report["extended_chunks_count"] = stats.extended_chunks_count;
+		report["extended_chunks_size"] = stats.extended_chunks_size;
+	}
+
+protected:
+	std::string name;
+	dataplane::memory_manager* memory_manager;
+	tSocketId socket_id;
+
+	object_type::stats_t stats;
+
+public:
+	object_type* pointer;
+};
+
+//
+
+template<typename key_t,
+         uint32_t chunk_size,
+         unsigned int valid_bit_offset = 0,
+         hash_function_t<key_t> calculate_hash = calculate_hash_crc<key_t>>
+class updater_hashtable_mod_id32
+{
+public:
+	using object_type = hashtable_mod_id32_dynamic<key_t, chunk_size, valid_bit_offset, calculate_hash>;
+
+	updater_hashtable_mod_id32(const char* name,
+	                           dataplane::memory_manager* memory_manager,
+	                           const tSocketId socket_id) :
+	        name(name),
+	        memory_manager(memory_manager),
+	        socket_id(socket_id),
+	        pointer(nullptr)
+	{
+	}
+
+	eResult update(const std::vector<std::tuple<key_t, uint32_t>>& values)
+	{
+		stats.pairs_size = upper_power_of_two(std::max(object_type::pairs_size_min,
+		                                               (uint32_t)(4ull * values.size())));
+
+		for (;;)
+		{
+			pointer = memory_manager->create<object_type>(name.data(),
+			                                              socket_id,
+			                                              object_type::calculate_sizeof(stats.pairs_size),
+			                                              stats.pairs_size);
+			if (pointer == nullptr)
+			{
+				return eResult::errorAllocatingMemory;
+			}
+
+			eResult result = pointer->fill(stats, values);
+			if (result != eResult::success)
```

Comment:
Do we really want to double the size of a hashtable in case of ANY error?

Reply:
`fill` can return success, or not enough memory